### PR TITLE
MIXEDARCH-352: Manage the lifecycle of the pod placement operand

### LINKED ISSUE #12: Implement  Operator and controllers as distinct entities with the operator-controller model
Currently the operator is implemented just as a p-o-c style controller to validate the functionality of adding node affinities on scheduling gated pods. But this is actually just the pod placement controller. We need the operator to manage various controllers such as:

- PodPlacement controller (which is built in to the operator today)
- PodPlacementConfig controller - managing and reconciling the config CR
-  Webhook controller - managing the webhook deployments

Other potential controllers to implement:
-  ICSP/IDMS controller - controller to reconcile and store/update ICSP/IDMS config
-  ImageRegistryConfig controller - update registry configuration and update policies.json

In addition to this, the PodPlacementController should run as an active-standby deployment using leader election. The webhook will be implemented as a replicaset (3 replicas ideally)  to intercept pods and gate them as per config.

This would involve the operator/controller each having their own deployment specs with command line options for each controller to be associated with the respective deployments

- Code would have to be restructured to have an operator package and a package for each controller
- Have deployment yamls for the operator and each controller

### PR DESCRIPTION
**Pull request by aleskandro**

### CODE REVIEW

File: apis/multiarch/v1alpha1/podplacementconfig_types.go [hunk 1]
```diff
@@ -30,6 +30,21 @@ const (
 	LogVerbosityLevelTraceAll LogVerbosityLevel = "TraceAll"
 )
 
+func (verbosity LogVerbosityLevel) ToZapLevelInt() int {
```
Comment (Prashanth684) [inline apis/multiarch/v1alpha1/podplacementconfig_types.go]: instead, can we use a map here? that maps the verbosity level to the int and use it directly

File: apis/multiarch/v1alpha1/podplacementconfig_types.go [hunk 2]
```diff
@@ -30,6 +30,21 @@ const (
...
+	switch verbosity {
+	case LogVerbosityLevelNormal:
+		return 3
```
Comment (Prashanth684) [inline apis/multiarch/v1alpha1/podplacementconfig_types.go]: even if it is small, i would suggest moving logging stuff to a separate helper file for improved readability

File: controllers/operator/objects.go [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+								"--v=0",
+							},
+							Ports: []corev1.ContainerPort{
+								{
+									ContainerPort: 8443,
+									Protocol:      corev1.ProtocolTCP,
+									Name:          "https",
+								},
+							},
+							Resources: corev1.ResourceRequirements{
```
Comment (Prashanth684) [inline controllers/operator/objects.go:222]: for future considerations, cpu and memory level would need to be tuneable

File: controllers/operator/objects.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/openshift/multiarch-manager-operator/apis/multiarch/v1alpha1"
+	"github.com/openshift/multiarch-manager-operator/pkg/utils"
+)
+
+func buildMutatingWebhookConfiguration(podPlacementConfig *v1alpha1.PodPlacementConfig) *admissionv1.MutatingWebhookConfiguration {
```
Comment (Prashanth684) [inline controllers/operator/objects.go:17]: is it too much overkill to use the[ deploymentcontroller](https://github.com/openshift/library-go/blob/master/pkg/operator/deploymentcontroller/deployment_controller.go) in library-go? it gives us more fine grained control that we might not need right now but maybe in the future

Comment (Prashanth684) [reply controllers/operator/objects.go:17]: we could use optional [hook functions](https://github.com/openshift/library-go/blob/master/pkg/operator/deploymentcontroller/deployment_controller.go#L59) to validate each part of the deployment

File: controllers/operator/objects.go (hunk 3 above)
Comment (aleskandro) [reply controllers/operator/objects.go:222]: Yeah, I had the same thought while writing this. I opened #22 to remember discussing this.

File: apis/multiarch/v1alpha1/podplacementconfig_types.go (hunk 1 above)
Comment (aleskandro) [reply apis/multiarch/v1alpha1/podplacementconfig_types.go]: Done. For now, I'm storing the map within the function to make the map itself 'constant/read-only'. Users requiring this conversion should still call a method in the type IMHO to follow the principle of encapsulation.

File: apis/multiarch/v1alpha1/podplacementconfig_types.go (hunk 2 above)
Comment (aleskandro) [reply apis/multiarch/v1alpha1/podplacementconfig_types.go]: done

File: controllers/operator/objects.go (hunk 4 above)
Comment (aleskandro) [reply controllers/operator/objects.go:17]: I also see this as useful as it helps defining the FSM for the ordered uninstallation in the future. Should we consider this as a follow-up and give priority to other stuff for now?

Comment (Prashanth684) [reply controllers/operator/objects.go:17]: let's follow up in a separate ticket.

Comment (aleskandro) [reply controllers/operator/objects.go:17]: #23

**Pull request merged**